staging/src/k8s.io/client-go/tools/cache/controller_test.go

```diff
@@ -893,3 +907,186 @@ func (m *mockTxnStore) Transaction(txns ...Transaction) *TransactionError {
 	m.succeedCount = len(txns)
 	return nil
 }
+
+func TestReplaceEvents(t *testing.T) {
+	_, ctx := ktesting.NewTestContext(t)
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	for _, atomic := range []bool{true, false} {
+		t.Run(fmt.Sprintf("Atomic=%v", atomic), func(t *testing.T) {
+			clientgofeaturetesting.SetFeatureDuringTest(t, clientgofeaturegate.AtomicFIFO, atomic)
+			source := fcache.NewFakeControllerSource()
+			store := NewStore(DeletionHandlingMetaNamespaceKeyFunc)
+			t.Cleanup(func() {
+				source.Shutdown()
+			})
+
+			recorder := &eventRecorder{}
+			c := newInformer(store, InformerOptions{
+				ListerWatcher: source,
+				ObjectType:    &v1.Pod{},
+				ResyncPeriod:  0,
+				Handler:       recorder,
+			})
+			go c.RunWithContext(ctx)
+			if !WaitForCacheSync(ctx.Done(), c.HasSynced) {
+				t.Fatal("Timed out waiting for cache sync")
+			}
+			testReplaceEvents(t, ctx, source, recorder, store)
+		})
+	}
+}
+
+func testReplaceEvents(t *testing.T, ctx context.Context, source *fcache.FakeControllerSource, m *eventRecorder, store Store) {
+	// Set up 3 pods initially
+	pKeep := &v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod-keep", Namespace: "default"}}
+	pMod := &v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod-mod", Namespace: "default", Labels: map[string]string{"ver": "1"}}}
+	pDel := &v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod-del", Namespace: "default"}}
+
+	source.Add(pKeep)
+	source.Add(pMod)
+	source.Add(pDel)
+
+	// Wait for all 3 events to appear
+	if err := wait.PollUntilContextTimeout(ctx, 10*time.Millisecond, 2*time.Second, true, func(_ context.Context) (bool, error) {
```

Comment:
To avoid calling lock on the separate data structure, maybe we could add `WaitForEventCount` method to recorder?